src/examples/TDXContainers.sol

```diff
@@ -0,0 +1,189 @@
+        /* TODO: verify measurements are as expected! */
+        /* TODO: require(_quoteMeasurementHash == _expectedMeasurement); */
+
+        bytes32 hostPubkeyHash;
+        assembly {
+            hostPubkeyHash := mload(add(rtmr3, 32))
+        }
+        require(hostPubkeyHash == keccak256(_hostPubkey));
+
+        keymgr.Suave().verifyTDXDCAPQuote(quote, pckCertPem, pckCrlPem, tcbInfo, qeIdentity);
+
+        return  (_quoteMeasurementHash, _hostPubkey, keymgr.attest(keccak256(abi.encodePacked("hPub", _hostPubkey, "qM", _quoteMeasurementHash))));
+    }
+
+    /* Offchain. Decrypts data using keymgr's key and encrypts it to host's pubkey */
+    function reencryptToHostKey(bytes memory ciphertext /* TODO: should we also require a specific quote measurement? */) external returns (bytes memory) {
+        bytes memory plaintext = PKE.decrypt(keymgr.derivedPriv(), ciphertext);
+        return PKE.encrypt(hostPubkey, keymgr.Suave().localRandom(), plaintext);
+    }
+
+    event NewContainerRequested(bytes32, ContainerConfig);
+    function submit(ContainerConfig calldata container) external {
+        bytes32 id = keccak256(abi.encode(container));
+        pendingContainers[id] = container;
+        emit NewContainerRequested(id, container);
+    }
+
+    function approve(bytes32 id) external {
+        require(msg.sender == owner);
+
+        ContainerConfig memory container = pendingContainers[id];
+        if (container.imageHash == bytes32(0)) revert("unknown container");
+
+        approvedContainers[id] = container;
+    }
+
+    event StartContainerRequested(bytes32, ContainerConfig);
+    function start(bytes32 id) external {
+        bytes32 runtimeId = keccak256(abi.encodePacked(id, msg.sender));
+        ContainerConfig memory container = runningContainers[runtimeId];
+        if (container.imageHash != bytes32(0)) {
+            emit StartContainerRequested(runtimeId, container);
+            return;
+        }
+
+        container = approvedContainers[id];
+        if (container.imageHash != bytes32(0)) {
+            emit StartContainerRequested(runtimeId, container);
+            runningContainers[runtimeId] = container;
+            return;
+        }
+
+        revert("unknown container");
+    }
+
+    event StopContainerRequested(bytes32, ContainerConfig);
+    function stop(bytes32 id) external {
+        bytes32 runtimeId = keccak256(abi.encodePacked(id, msg.sender));
+
+        ContainerConfig memory container = runningContainers[runtimeId];
+        if (container.imageHash != bytes32(0)) {
+            emit StopContainerRequested(runtimeId, container);
+            delete runningContainers[runtimeId];
+            return;
+        }
+
+        container = approvedContainers[id];
+        if (container.imageHash != bytes32(0)) {
+            emit StopContainerRequested(runtimeId, container);
+            return;
+        }
+
+        revert("unknown container");
+    }
+
+    function heartbeat() external {
```

Comment:
this is cool, i definitely think it's useful to have a heartbeat as a way of getting affirmative updates that a container has made progress and is up. Hard to tell if it's actually dead or pretending to be offline